pkg/lib/proxy/syncer.go

```diff
@@ -0,0 +1,87 @@
+package proxy
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/openshift/client-go/config/informers/externalversions"
+
+	apiconfigv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned"
+	configv1 "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/discovery"
+)
+
+const (
+	// This is the cluster level global proxy object name.
+	globalProxyName = "cluster"
+
+	// default sync interval
+	defaultSyncInterval = 30 * time.Minute
+)
+
+// NewSyncer returns informer and sync functions to enable watch of Proxy type.
+func NewSyncer(logger *logrus.Logger, client configv1client.Interface, discovery discovery.DiscoveryInterface) (proxyInformer configv1.ProxyInformer, syncer *Syncer, querier Querier, err error) {
+	factory := externalversions.NewSharedInformerFactoryWithOptions(client, defaultSyncInterval)
+	proxyInformer = factory.Config().V1().Proxies()
+	s := &Syncer{
+		logger: logger,
+		lister: proxyInformer.Lister(),
+	}
+
+	syncer = s
+	querier = s
+	return
+}
+
+// Syncer deals with watching proxy type(s) on the cluster and let the caller
+// query for cluster scoped proxy objects.
+type Syncer struct {
+	logger *logrus.Logger
+	lister listers.ProxyLister
+}
+
+// QueryProxyConfig queries the global cluster level proxy object and then
+// returns the proxy env variable(s) to the user.
+func (w *Syncer) QueryProxyConfig() (proxy []corev1.EnvVar, err error) {
+	global, getErr := w.lister.Get(globalProxyName)
+	if getErr != nil {
+		if !k8serrors.IsNotFound(getErr) {
+			err = getErr
+			return
+		}
+
+		err = fmt.Errorf("global Proxy configuration not defined in '%s' - %v", globalProxyName, getErr)
```

Comment:
If an openshift cluster does not have the `cluster` proxy object then deployment initialization will fail and the operator install will fail as well.

Reply:
I don't think it's required that an openshift cluster have a cluster proxy object, right? Just that if it is there, we should respect it?

Reply:
Yes, this is much safer. And for a scenario where there was a `cluster` object and then it gets removed later - we will not find a `cluster` proxy object and generate a pod spec without the proxy env vars. This will cause a mismatch of the desired and observed and eventually update deployment(s) (originally created with the proxy env variable(s) )  to not have them any longer, correct?

Reply:
Yep, that sounds right